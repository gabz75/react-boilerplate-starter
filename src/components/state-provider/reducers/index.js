import notificationReducer, {
  initialState as notificationsInitialState,
} from './notifications';

export const initialState = {
  notifications: notificationsInitialState,
};

export const reducers = {
  notifications: notificationReducer,
};
