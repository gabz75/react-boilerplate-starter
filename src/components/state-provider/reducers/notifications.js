export const initialState = {
  data: [],
};

const ADD_NOTIFICATION = 'ADD_NOTIFICATION';
const REMOVE_NOTIFICATION = 'REMOVE_NOTIFICATION';

export function addNotification(notification) {
  return {
    reducer: 'notifications',
    type: ADD_NOTIFICATION,
    payload: {
      notification,
    },
  };
}

export function removeNotification(notification) {
  return {
    reducer: 'notifications',
    type: REMOVE_NOTIFICATION,
    payload: {
      notification,
    },
  };
}

function reduceAddNotification(state, action = {}) {
  const { notification } = action.payload;
  const { data } = state.notifications;
  data.push(notification);
  return { ...state, data: [...data] };
}

function reduceRemoveNotification(state, action = {}) {
  const { notification } = action.payload;
  const { data } = state.notifications;
  const newData = data.filter((n) => n.id !== notification.id);
  return { ...state, notifications: { data: newData } };
}

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case ADD_NOTIFICATION:
      return reduceAddNotification(state, action);
    case REMOVE_NOTIFICATION:
      return reduceRemoveNotification(state, action);
    default:
      return state;
  }
}
