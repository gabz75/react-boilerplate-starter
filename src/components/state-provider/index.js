import React, { createContext, useReducer, useContext } from 'react';
import PropTypes from 'prop-types';

import { initialState, reducers } from './reducers';

function reducer(state = initialState, action = {}) {
  if (reducers[action.reducer]) {
    return reducers[action.reducer](state, action);
  }

  return state;
}

const StoreContext = createContext(initialState);

/**
 * Main <StateProvide> Provider/Consumer Component
 * @param {[type]} options.children [description]
 */
export default function StateProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);
  const value = { state, dispatch };

  return <StoreContext.Provider value={value}>{children}</StoreContext.Provider>;
}

StateProvider.propTypes = {
  children: PropTypes.element.isRequired,
};

/**
 * Hook to use the store and grab the { state, dispatch } object/method.
 * @return {Objet]} state + dispatch
 */
export const useStore = () => useContext(StoreContext);
