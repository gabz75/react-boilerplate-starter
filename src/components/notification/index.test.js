import React from 'react';
import { render } from '@testing-library/react';

import { REMOVE_NOTIFICATION } from 'components/state-provider/notifications';
// import { mockUseStore } from 'tests/mock-use-store';
import { mockUseDispatch } from 'tests/mock-use-dispatch';
import { mockUseSelector } from 'tests/mock-use-selector';

import Notification, { NotificationDismissButton, NotificationOverlay } from './index';

import {
  LEVEL_INFO,
  LEVEL_ERROR,
  LEVEL_WARNING,
} from './style';

const defaultNotification = {
  id: '1',
  content: 'mock-content',
  dismissible: true,
  level: LEVEL_INFO,
};

describe('NotificationDismissButton', () => {
  it('renders', () => {
    const { asFragment } = render(<NotificationDismissButton onClick={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('executes a callback', () => {
    const clickHandler = jest.fn();
    const { getByTestId } = render(<NotificationDismissButton onClick={clickHandler} />);

    getByTestId('dismiss-button').click();

    expect(clickHandler).toHaveBeenCalled();
  });

  it('throws an error when onClick is missing', () => {
    const spy = jest.spyOn(global.console, 'error').mockImplementation(() => {});
    render(<NotificationDismissButton />);

    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });
});

describe('NotificationOverlay', () => {
  it('renders', () => {
    const { asFragment } = render(<NotificationOverlay notification={defaultNotification} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders an error', () => {
    const notification = { ...defaultNotification, level: LEVEL_ERROR };
    const { asFragment } = render(<NotificationOverlay notification={notification} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a warning', () => {
    const notification = { ...defaultNotification, level: LEVEL_WARNING };
    const { asFragment } = render(<NotificationOverlay notification={notification} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a non-dismissible notification', () => {
    const notification = { ...defaultNotification, dismissible: false };
    const { queryByTestId } = render(<NotificationOverlay notification={notification} />);
    expect(queryByTestId('dismiss-button')).not.toBeInTheDocument();
  });

  it('renders a dismissible notification by default', () => {
    const { queryByTestId } = render(<NotificationOverlay notification={defaultNotification} />);
    expect(queryByTestId('dismiss-button')).toBeInTheDocument();
  });

  it('dismiss dispatches a REMOVE_NOTIFICATION action', () => {
    const { dispatch, restore } = mockUseDispatch();
    const { getByTestId } = render(<NotificationOverlay notification={defaultNotification} />);
    getByTestId('dismiss-button').click();

    expect(dispatch).toHaveBeenCalledWith({
      type: REMOVE_NOTIFICATION,
      payload: { notification: defaultNotification },
    });

    restore();
  });

  it('given dismissIn, dispatches a REMOVE_NOTIFICATION action', () => {
    jest.useFakeTimers();

    const { dispatch, restore } = mockUseDispatch();
    const notification = { ...defaultNotification, dismissIn: 5000 };

    render(<NotificationOverlay notification={notification} />);

    jest.advanceTimersByTime(4999);

    expect(dispatch).not.toHaveBeenCalled();

    jest.advanceTimersByTime(1);

    expect(dispatch).toHaveBeenCalled();

    restore();
  });
});

describe('Notification', () => {
  it('renders an empty notification list', () => {
    const { restore } = mockUseSelector({ notifications: { data: [] } });
    const { asFragment, queryByTestId, queryAllByTestId } = render(<Notification />);

    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId('notification-wrapper')).toBeInTheDocument();
    expect(queryAllByTestId('dismiss-button')).toHaveLength(0);

    restore();
  });

  it('renders a notification list', () => {
    const { restore } = mockUseSelector({
      notifications: {
        data: [
          { ...defaultNotification },
          { ...defaultNotification, id: '2' },
        ],
      },
    });
    const { asFragment, queryByTestId, queryAllByTestId } = render(<Notification />);

    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId('notification-wrapper')).toBeInTheDocument();
    expect(queryAllByTestId('dismiss-button')).not.toHaveLength(0);

    restore();
  });
});
