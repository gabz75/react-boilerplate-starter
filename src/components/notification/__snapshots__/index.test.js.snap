// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders a notification list 1`] = `
<DocumentFragment>
  <div
    class="sc-AykKF bTAboA"
    data-testid="notification-wrapper"
  >
    <div
      class="sc-AykKG dchsrd"
    >
      <button
        class="sc-AykKE sc-AykKH gMOKrM"
        data-testid="dismiss-button"
      >
        <svg
          aria-hidden="true"
          class="octicon"
          height="16"
          role="img"
          style="display: inline-block; fill: currentColor; user-select: none; vertical-align: text-bottom;"
          viewBox="0 0 12 16"
          width="12"
        >
          <path
            d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      mock-content
    </div>
    <div
      class="sc-AykKG dchsrd"
    >
      <button
        class="sc-AykKE sc-AykKH gMOKrM"
        data-testid="dismiss-button"
      >
        <svg
          aria-hidden="true"
          class="octicon"
          height="16"
          role="img"
          style="display: inline-block; fill: currentColor; user-select: none; vertical-align: text-bottom;"
          viewBox="0 0 12 16"
          width="12"
        >
          <path
            d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      mock-content
    </div>
  </div>
</DocumentFragment>
`;

exports[`Notification renders an empty notification list 1`] = `
<DocumentFragment>
  <div
    class="sc-AykKF bTAboA"
    data-testid="notification-wrapper"
  />
</DocumentFragment>
`;

exports[`NotificationDismissButton renders 1`] = `
<DocumentFragment>
  <button
    class="sc-AykKE sc-AykKH gMOKrM"
    data-testid="dismiss-button"
  >
    <svg
      aria-hidden="true"
      class="octicon"
      height="16"
      role="img"
      style="display: inline-block; fill: currentColor; user-select: none; vertical-align: text-bottom;"
      viewBox="0 0 12 16"
      width="12"
    >
      <path
        d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`NotificationOverlay renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AykKG dchsrd"
  >
    <button
      class="sc-AykKE sc-AykKH gMOKrM"
      data-testid="dismiss-button"
    >
      <svg
        aria-hidden="true"
        class="octicon"
        height="16"
        role="img"
        style="display: inline-block; fill: currentColor; user-select: none; vertical-align: text-bottom;"
        viewBox="0 0 12 16"
        width="12"
      >
        <path
          d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    mock-content
  </div>
</DocumentFragment>
`;

exports[`NotificationOverlay renders a warning 1`] = `
<DocumentFragment>
  <div
    class="sc-AykKG goAyQy"
  >
    <button
      class="sc-AykKE sc-AykKH gMOKrM"
      data-testid="dismiss-button"
    >
      <svg
        aria-hidden="true"
        class="octicon"
        height="16"
        role="img"
        style="display: inline-block; fill: currentColor; user-select: none; vertical-align: text-bottom;"
        viewBox="0 0 12 16"
        width="12"
      >
        <path
          d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    mock-content
  </div>
</DocumentFragment>
`;

exports[`NotificationOverlay renders an error 1`] = `
<DocumentFragment>
  <div
    class="sc-AykKG eWROoU"
  >
    <button
      class="sc-AykKE sc-AykKH gMOKrM"
      data-testid="dismiss-button"
    >
      <svg
        aria-hidden="true"
        class="octicon"
        height="16"
        role="img"
        style="display: inline-block; fill: currentColor; user-select: none; vertical-align: text-bottom;"
        viewBox="0 0 12 16"
        width="12"
      >
        <path
          d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    mock-content
  </div>
</DocumentFragment>
`;
